packages/inter-protocol/src/auction/auctionBook.js

```diff
@@ -454,38 +456,59 @@ export const prepareAuctionBook = (baggage, zcf, makeRecorderKit) => {
           });
           return state.bookDataKit.recorder.write(bookData);
         },
-        observeQuoteNotifier() {
+        // Ensure that there is an observer monitoring the quoteNotifier. We
+        // assume that all failure modes for quoteNotifier eventually lead to
+        // fail or finish.
+        ensureQuoteNotifierObserved() {
           const { state, facets } = this;
           const { collateralBrand, bidBrand, priceAuthority } = state;
 
+          if (observedBrands.has(collateralBrand)) {
+            return;
+          }
+          observedBrands.add(collateralBrand);
           trace('observing');
 
-          const quoteNotifier = E(priceAuthority).makeQuoteNotifier(
+          const quoteNotifierP = E(priceAuthority).makeQuoteNotifier(
             AmountMath.make(collateralBrand, QUOTE_SCALE),
             bidBrand,
           );
-          void observeNotifier(quoteNotifier, {
-            updateState: quote => {
-              trace(
-                `BOOK notifier ${priceFrom(quote).numerator.value}/${
-                  priceFrom(quote).denominator.value
-                }`,
-              );
-              state.updatingOracleQuote = priceFrom(quote);
-            },
-            fail: reason => {
-              trace(`Failure from quoteNotifier (${reason}) setting to null`);
-              // lack of quote will trigger restart
-              state.updatingOracleQuote = null;
-            },
-            finish: done => {
-              trace(
-                `quoteNotifier invoked finish(${done}). setting quote to null`,
-              );
-              // lack of quote will trigger restart
+
+          void E.when(
+            quoteNotifierP,
+            quoteNotifier =>
+              observeNotifier(quoteNotifier, {
+                updateState: quote => {
+                  trace(
+                    `BOOK notifier ${priceFrom(quote).numerator.value}/${
+                      priceFrom(quote).denominator.value
+                    }`,
+                  );
+                  state.updatingOracleQuote = priceFrom(quote);
+                },
+                fail: reason => {
+                  trace(
+                    `Failure from quoteNotifier (${reason}) setting to null`,
+                  );
+                  // lack of quote will trigger restart
+                  state.updatingOracleQuote = null;
+                  observedBrands.delete(collateralBrand);
+                },
+                finish: done => {
+                  trace(
+                    `quoteNotifier invoked finish(${done}). setting quote to null`,
+                  );
+                  // lack of quote will trigger restart
+                  state.updatingOracleQuote = null;
+                  observedBrands.delete(collateralBrand);
+                },
+              }),
+            e => {
+              trace('makeQuoteNotifier failed, resetting', e);
               state.updatingOracleQuote = null;
```

Comment:
Why is it necessary to reset the state here?

Reply:
It's safe because we'll check for a notifier again at the next auction.  It's a good idea because we can't count on the notifier continuing to produce events if there's an error at this level.

Reply:
I just don't quite understand in which case this isn't already `null`. This is an error making the notifier, so afaiu we would never have updated this state in the first place?

Reply:
We didn't explore it deeply, but I think if a previous version was happily updating the state, then it upgrades, then anything which samples the price at the beginning of the new incarnation (before any new updates happen) will get the old price. If an auction then starts and initiates the observer process, but fails (maybe because the price feed vat is getting upgraded too?), then we'll hit this case, clear the ephemeral flag (so the *next* auction can initiate a new one), and also clear the durable `state.updatingOracleQuote`, which at least will tell clients to not rely on the old quote.

I suspect that's a bit weird. Not sure if it's a problem though.

Reply:
On line 469, we add the brand to the set, and on line 472, we attempt to create the notifier. On line 477, we await the notifier. If the creation promise fails, we end  up here, and want to ensure the brand is removed from the set, or we'll never try again to create a notifier.

Reply:
> ensure the brand is removed from the set, or we'll never try again to create a notifier.

I have no problem with the ephemeral set being reset on error. My concern is with how the durable `updatingOracleQuote` state is managed. Resetting it at this state doesn't seem correct to me, and might be an indication something else is amiss.

Brian mentions a case where this durable state may not be `null`: after an upgrade. I suppose the product question is whether we want to maintain or not the old "updating" quote while we attempt to restart the observer.

But beyond that question, the problem is that the `ensureQuoteNotifierObserved` call is gated on `!updatingOracleQuote` in the first place, so really it seems like restarting the vat wouldn't trigger the observer creation anyway because that state wouldn't be reset by the upgrade, and the contract would think it still has a notifier.

IMO, we need to switch the call to `ensureQuoteNotifierObserved` to be unconditional, return `true` if the observer is in the Set, and if not immediately reset the `updatingOracleQuote` (if we want to avoid pretending we have an updating quote), start the notifier, and return false so that `captureOraclePriceForRound` can continue bailing out early. In that case `updatingOracleQuote` wouldn't need to be reset if we fail to make the observer since we'd be guaranteed for it to already be `null` (again if we don't want to pretend we have an updating quote while re-establishing the observer).

Regarding the early bail out of `captureOraclePriceForRound`, I'm a little surprised that we would skip capturing a price for round when the observer isn't yet setup instead of delaying the capture, but that's a different topic.

Reply:
Oh, yeah, if the auctionBook has a price (in durable `state.updatingOracleQuote`), then gets restarted/upgraded, the new incarnation won't start a new observer at startup (because nothing special happens at restart, it's not like we walk through all the pre-existing brands and do something for each one), and when the auction timer fires and it calls `captureOraclePriceForRound()`, that will see the old price and assume it's ok, and won't start a replacement observer then either.

I agree with @mhofman that the `captureOraclePriceForRound()` needs to check the ephemeral set instead of the durable state, or simply always call `ensureQuoteNotifierObserved()` and rely on its internal check to prevent duplicates.